cedar-lean/Cedar/Spec/PartialExpr.lean

```diff
@@ -0,0 +1,212 @@
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+
+      https://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+-/
+
+import Cedar.Data
+import Cedar.Spec.Expr
+import Cedar.Thm.Data.Map -- necessary for a termination argument
+import Cedar.Thm.Termination.Expr
+import Cedar.Thm.Termination.Value
+
+/-! This file defines abstract syntax for Cedar expressions. -/
+
+namespace Cedar.Spec
+
+open Cedar.Data
+
+/--
+  Identical to `Expr` except that it has an `unknown` case, and the recursive
+  elements are also all `PartialExpr` instead of `Expr`
+-/
+inductive PartialExpr where
+  | lit (p : Prim)
+  | var (v : Var)
+  | ite (cond : PartialExpr) (thenExpr : PartialExpr) (elseExpr : PartialExpr)
+  | and (a : PartialExpr) (b : PartialExpr)
+  | or (a : PartialExpr) (b : PartialExpr)
+  | unaryApp (op : UnaryOp) (expr : PartialExpr)
+  | binaryApp (op : BinaryOp) (a : PartialExpr) (b : PartialExpr)
+  | getAttr (expr : PartialExpr) (attr : Attr)
+  | hasAttr (expr : PartialExpr) (attr : Attr)
+  | set (ls : List PartialExpr)
+  | record (map : List (Attr × PartialExpr))
+  | call (xfn : ExtFun) (args : List PartialExpr)
+  | unknown (name : String)
+
+deriving instance Repr, Inhabited for PartialExpr
+
+def Value.asPartialExpr (v : Value) : PartialExpr :=
+  match v with
+  | .prim p => .lit p
+  | .set s => .set (s.elts.map₁ λ ⟨v, h₁⟩ =>
+      have := Value.set_termination v s h₁
+      v.asPartialExpr)
+  | .record m => .record (m.kvs.map₁ λ ⟨(k, v), h₁⟩ =>
+      have := Value.record_termination v m (Map.in_kvs_values h₁)
+      (k, v.asPartialExpr))
+  | .ext (.decimal d) => .call ExtFun.decimal [PartialExpr.lit (.string d.unParse)]
+  | .ext (.ipaddr ip) => .call ExtFun.ip [PartialExpr.lit (.string (Cedar.Spec.Ext.IPAddr.unParse ip))]
+termination_by v.size
+
+/--
+  A version of `PartialExpr`, but only allows "restricted expressions" -- no
+  vars, no expressions that require entity data to evaluate, no operators at all,
+  just literals, unknowns, extension values, and sets/records of those things
+-/
+inductive RestrictedPartialExpr where
+  | lit (p : Prim)
+  | set (ls : List RestrictedPartialExpr)
+  | record (map : List (Attr × RestrictedPartialExpr))
+  | call (xfn : ExtFun) (args : List Value) -- this requires that all arguments to extension functions in RestrictedPartialExpr are concrete. TODO do we need to relax this?
+  | unknown (name : String)
+
+deriving instance Repr, Inhabited for RestrictedPartialExpr
+
+def Value.asRestrictedPartialExpr (v : Value) : RestrictedPartialExpr :=
+  match v with
+  | .prim p => .lit p
+  | .set s => .set (s.elts.map₁ λ ⟨v, h₁⟩ =>
```

Comment:
I'm hoping we can simplify this, to get rid of the calls to the termination lemmas.  I'll play it with a bit offline to see what we can do.

Reply:
Turns out we can avoid these explicit proofs :)

Here is the one for `Expr.asPartialExpr`: https://github.com/cedar-policy/cedar-spec/blob/7d05c98c753b1208aa803894842180d83b83e3a9/cedar-lean/Cedar/Partial/Expr.lean#L167

Others should be similar.

The key is to use `.attach₂` to map over records in a way that ensures termination.

Reply:
Thanks for this, this is helpful